pkg/block/block.go

```diff
@@ -0,0 +1,715 @@
+	for _, v := range a {
+		key, ok := extractArrayEntryKey(v, block.Key)
+		if !ok {
+			continue
+		}
+		// if there is no TypeKey, we'll use "" to indicate that
+		entryType, ok := extractArrayEntryType(v, block.TypeKey)
+		if !ok {
+			entryType = ""
+		}
+		entries[key] = entry{
+			A:     v.(map[string]any),
+			AType: entryType,
+		}
+	}
+	for _, v := range b {
+		key, ok := extractArrayEntryKey(v, block.Key)
+		if !ok {
+			continue
+		}
+		ty, ok := extractArrayEntryType(v, block.TypeKey)
+		if !ok {
+			ty = ""
+		}
+		e := entries[key]
+		e.B = v.(map[string]any)
+		e.BType = ty
+		entries[key] = e
+	}
+
+	var patches []Patch
+	for k, e := range entries {
+		if e.B == nil {
+			// array element was deleted
+			patches = append(patches, Patch{
+				Path:    []PathSegment{{ArrayKey: block.Key, ArrayElem: k}},
+				Deleted: true,
+			})
+		} else if e.AType != e.BType {
+			// array element changed types and therefore block list, not directly comparable, replace the entire element
+			// (this also covers the case where a new element was added)
+			patches = append(patches, Patch{
+				Path:  []PathSegment{{ArrayKey: block.Key, ArrayElem: k}},
+				Value: e.B,
+			})
+		} else {
+			blocks, ok := block.BlocksByType[e.BType]
+			if !ok {
+				// fall back to the global blocks
+				// this case is used when SplitByKey isn't used
+				blocks = block.Blocks
+			}
+
+			subpatches := diff(e.A, e.B, Block{Blocks: blocks})
+			prefixPatches(subpatches, []PathSegment{{ArrayKey: block.Key, ArrayElem: k}})
+			patches = append(patches, subpatches...)
+		}
+	}
+
+	return patches
+}
+
+// extracts the logical key from an object which is used to identify it within an array
+// if a key is returned, it is guaranteed to be comparable
+func extractArrayEntryKey(v any, key string) (any, bool) {
+	m, ok := v.(map[string]any)
+	if !ok {
+		return nil, false
+	}
+	keyValue, ok := m[key]
+	if !ok || !reflect.ValueOf(keyValue).Comparable() {
+		return nil, false
+	}
+	return keyValue, true
+}
+
+func extractArrayEntryType(v any, typeKey string) (string, bool) {
+	m, ok := v.(map[string]any)
+	if !ok {
+		return "", false
+	}
+	entryType, ok := m[typeKey]
+	if !ok {
+		return "", false
+	}
+	entryTypeStr, ok := entryType.(string)
+	return entryTypeStr, ok
+}
+
+// deep equality check for supporting the following nested types:
+// - map[string]any
+// - []any
+// - comparable types
+func equal(a, b any) bool {
+	switch a := a.(type) {
+	case map[string]any:
+		b, ok := b.(map[string]any)
+		if !ok {
+			return false
+		}
+		if !equalKeys(a, b) {
+			return false
+		}
+		for k, v := range a {
+			if !equal(v, b[k]) {
+				return false
+			}
+		}
+		return true
+
+	case []any:
+		b, ok := b.([]any)
+		if !ok {
+			return false
+		}
+		if len(a) != len(b) {
+			return false
+		}
+		for i := range a {
+			if !equal(a[i], b[i]) {
+				return false
+			}
+		}
+		return true
+
+	default:
+		return a == b
```

Comment:
I'm not sure if this can actually happen, but just to be aware that comparing two `any` types that actually aren't comparable will result in a runtime panic

Reply:
This can't happen because of the round-trip through json.Marshal/Unmarshal.
That will only output `map[string]any`, `[]any` (which are special-cased), `string`, `bool` and `float64` which are all comparable.

The comment indicates the permitted types. I'm happy to leave as-is because:

1. The logic of the program should prevent an invalid value appearing here
2. If there is somehow a bug which allows this, it'll panic and we'll know.
3. I can't think of a sensible behaviour for incomparable types (return `false`?)

Reply:
I'm happy to leave as is.